File: flang_modernizer/transform.py
from __future__ import annotations

import re
from pathlib import Path
from typing import List


DECLARATION_RE = re.compile(r"^\s*(integer|real|double\s+precision|complex|logical|character)\b", re.IGNORECASE)
STATEMENT_FUNCTION_RE = re.compile(r"^\s*([a-z_]\w*)\s*\(([^)]*)\)\s*=\s*(.+)$", re.IGNORECASE)


def apply_safe_transformations(source: Path, output_dir: Path) -> List[Path]:
    """Apply deliberately conservative demo transformations.

    These transformations are syntax-preserving cleanups intended to validate the
    advisor's low-risk recommendations. They do not attempt semantic rewrites for
    COMMON, EQUIVALENCE, ENTRY, or computed branches.
    """

    output_dir.mkdir(parents=True, exist_ok=True)
    changed: List[Path] = []
    paths = [source] if source.is_file() else sorted(p for p in source.rglob("*") if p.suffix.lower() in {".f", ".for", ".f77"})
    for path in paths:
        original = path.read_text(errors="replace").splitlines()
        rewritten = [_rewrite_fixed_line(line) for line in original]
        rewritten = _convert_simple_statement_functions(rewritten)
        target = output_dir / path.with_suffix(".f90").name
        target.write_text("\n".join(rewritten) + "\n")
        changed.append(target)
    return changed


def _rewrite_fixed_line(line: str) -> str:
    if not line:
        return line
    if line[:1].upper() in {"C", "*"}:
        return "!" + line[1:]
    if len(line) >= 6 and line[:5].strip().isdigit():
        return line[:5].strip() + " " + line[6:].rstrip()
    if len(line) >= 6 and line[5].strip():
        return "& " + line[6:].rstrip()
    return line.rstrip()


def _convert_simple_statement_functions(lines: List[str]) -> List[str]:
    result: List[str] = []
    declared: set[str] = set()
    pending_functions: List[str] = []
    in_routine = False

    for line in lines:
        stripped = line.strip()
        if re.match(r"^(program|subroutine|function)\b", stripped, re.IGNORECASE):
            in_routine = True
            declared = set()
            pending_functions = []

        if in_routine and DECLARATION_RE.match(stripped):
            declared.update(_declared_names(stripped))

        stmt_fn = STATEMENT_FUNCTION_RE.match(stripped) if in_routine else None
        if stmt_fn and stmt_fn.group(1).lower() not in declared:
            pending_functions.extend(_internal_function_lines(stmt_fn.group(1), stmt_fn.group(2), stmt_fn.group(3)))
            continue

        if in_routine and re.match(r"^end\b", stripped, re.IGNORECASE):
            if pending_functions:
                result.append("contains")
                result.extend(pending_functions)
            in_routine = False

        result.append(line)
    return result


def _declared_names(code: str) -> set[str]:
    _, _, tail = code.partition("::")
    if not tail:
        tail = re.sub(
            r"^\s*(integer|real|complex|logical|character|double\s+precision)"
            r"(?:\s*\*?\s*\d+|\s*\([^)]*\))?\s*",
            "",
            code,
            flags=re.IGNORECASE,
        )
    names = set()
    for part in tail.split(","):
        name = re.sub(r"\(.*", "", part).strip().lower()
        if re.match(r"^[a-z_]\w*$", name):
            names.add(name)
    return names


def _internal_function_lines(name: str, args: str, expression: str) -> List[str]:
    arg_names = [arg.strip() for arg in args.split(",") if arg.strip()]
    lines = [f"  real function {name}({', '.join(arg_names)})"]
    for arg in arg_names:
        lines.append(f"    real, intent(in) :: {arg}")
    lines.append(f"    {name} = {expression.strip()}")
    lines.append(f"  end function {name}")
    return lines

File: src/Transform.cpp
#include "Transform.hpp"

#include <algorithm>
#include <fstream>
#include <regex>
#include <set>
#include <sstream>

namespace {
std::string trim(const std::string &value) {
  auto start = value.find_first_not_of(" \t\r\n");
  if (start == std::string::npos) {
    return "";
  }
  auto end = value.find_last_not_of(" \t\r\n");
  return value.substr(start, end - start + 1);
}

std::string lower(std::string value) {
  std::transform(value.begin(), value.end(), value.begin(), [](unsigned char ch) { return std::tolower(ch); });
  return value;
}

std::vector<std::string> readLines(const std::filesystem::path &path) {
  std::ifstream input(path);
  std::vector<std::string> lines;
  std::string line;
  while (std::getline(input, line)) {
    lines.push_back(line);
  }
  return lines;
}

void writeLines(const std::filesystem::path &path, const std::vector<std::string> &lines) {
  std::ofstream output(path);
  for (const auto &line : lines) {
    output << line << "\n";
  }
}

bool isFixedSource(const std::filesystem::path &path) {
  auto ext = lower(path.extension().string());
  return ext == ".f" || ext == ".for" || ext == ".f77" || ext == ".ftn";
}

std::string rewriteFixedLine(const std::string &line) {
  if (line.empty()) {
    return line;
  }
  if (line[0] == 'C' || line[0] == 'c' || line[0] == '*') {
    return "!" + line.substr(1);
  }
  if (line.size() >= 6) {
    std::string label = trim(line.substr(0, 5));
    if (!label.empty() && label.find_first_not_of("0123456789") == std::string::npos) {
      return label + " " + line.substr(6);
    }
    if (!trim(line.substr(5, 1)).empty()) {
      return "& " + line.substr(6);
    }
  }
  return line;
}

std::set<std::string> declaredNames(const std::string &code) {
  std::string tail = code;
  auto pos = tail.find("::");
  if (pos != std::string::npos) {
    tail = tail.substr(pos + 2);
  } else {
    tail = std::regex_replace(tail,
                              std::regex(R"(^\s*(integer|real|complex|logical|character|double\s+precision)(?:\s*\*?\s*\d+|\s*\([^)]*\))?\s*)",
                                         std::regex::icase),
                              "");
  }
  std::set<std::string> names;
  std::stringstream stream(tail);
  std::string part;
  while (std::getline(stream, part, ',')) {
    auto paren = part.find('(');
    if (paren != std::string::npos) {
      part = part.substr(0, paren);
    }
    part = lower(trim(part));
    if (std::regex_match(part, std::regex(R"([a-z_]\w*)", std::regex::icase))) {
      names.insert(part);
    }
  }
  return names;
}

std::vector<std::string> internalFunctionLines(const std::string &name, const std::string &args,
                                               const std::string &expression) {
  std::vector<std::string> argNames;
  std::stringstream stream(args);
  std::string arg;
  while (std::getline(stream, arg, ',')) {
    arg = trim(arg);
    if (!arg.empty()) {
      argNames.push_back(arg);
    }
  }

  std::ostringstream signature;
  signature << "  real function " << name << "(";
  for (size_t i = 0; i < argNames.size(); ++i) {
    if (i > 0) {
      signature << ", ";
    }
    signature << argNames[i];
  }
  signature << ")";

  std::vector<std::string> lines{signature.str()};
  for (const auto &argName : argNames) {
    lines.push_back("    real, intent(in) :: " + argName);
  }
  lines.push_back("    " + name + " = " + trim(expression));
  lines.push_back("  end function " + name);
  return lines;
}

std::vector<std::string> convertSimpleStatementFunctions(const std::vector<std::string> &lines) {
  const std::regex routineRe(R"(^(program|subroutine|function)\b)", std::regex::icase);
  const std::regex endRe(R"(^end\b)", std::regex::icase);
  const std::regex declarationRe(R"(^\s*(integer|real|double\s+precision|complex|logical|character)\b)",
                                 std::regex::icase);
  const std::regex statementFunctionRe(R"(^\s*([a-z_]\w*)\s*\(([^)]*)\)\s*=\s*(.+)$)",
                                       std::regex::icase);

  std::vector<std::string> result;
  std::set<std::string> declared;
  std::vector<std::string> pendingFunctions;
  bool inRoutine = false;

  for (const auto &line : lines) {
    std::string stripped = trim(line);
    if (std::regex_search(stripped, routineRe)) {
      inRoutine = true;
      declared.clear();
      pendingFunctions.clear();
    }
    if (inRoutine && std::regex_search(stripped, declarationRe)) {
      auto names = declaredNames(stripped);
      declared.insert(names.begin(), names.end());
    }

    std::smatch match;
    if (inRoutine && std::regex_search(stripped, match, statementFunctionRe) && declared.count(lower(match[1].str())) == 0) {
      auto generated = internalFunctionLines(match[1].str(), match[2].str(), match[3].str());
      pendingFunctions.insert(pendingFunctions.end(), generated.begin(), generated.end());
      continue;
    }

    if (inRoutine && std::regex_search(stripped, endRe)) {
      if (!pendingFunctions.empty()) {
        result.push_back("contains");
        result.insert(result.end(), pendingFunctions.begin(), pendingFunctions.end());
      }
      inRoutine = false;
    }
    result.push_back(line);
  }
  return result;
}
} // namespace

std::vector<std::filesystem::path> applySafeTransformations(const std::filesystem::path &source,
                                                            const std::filesystem::path &outputDir) {
  std::filesystem::create_directories(outputDir);
  std::vector<std::filesystem::path> inputs;
  if (std::filesystem::is_regular_file(source)) {
    inputs.push_back(source);
  } else {
    for (const auto &entry : std::filesystem::recursive_directory_iterator(source)) {
      if (entry.is_regular_file() && isFixedSource(entry.path())) {
        inputs.push_back(entry.path());
      }
    }
  }
  std::sort(inputs.begin(), inputs.end());

  std::vector<std::filesystem::path> written;
  for (const auto &input : inputs) {
    auto lines = readLines(input);
    for (auto &line : lines) {
      line = rewriteFixedLine(line);
    }
    lines = convertSimpleStatementFunctions(lines);

    auto target = outputDir / input.filename();
    target.replace_extension(".f90");
    writeLines(target, lines);
    written.push_back(target);
  }
  return written;
}
